### DISCUSSION TITLE
`std::atomic<std::shared_ptr>::wait` does not seem to care about control block difference. Is this a bug?

### DISCUSSION THREAD

**Discussion by zwhfly** [microsoft/STL#3555] [Q&A] [answered]
This code seems to block forever under MSVC, but returns successfully under <del>gcc/clang</del>[libstdc++](https://godbolt.org/z/hG8e6oYKn).

```
#include <memory>
#include <atomic>


int main()
{
    auto sp1 = std::make_shared<char>();
    auto holder = [sp1] {};
    auto sp2 = std::make_shared<decltype(holder)>(holder);
    std::shared_ptr<char> sp3{ sp2, sp1.get() };

    std::atomic<std::shared_ptr<char>> asp{ sp1 };
    asp.wait(sp3);

    return 0;
}
```

It is said [here](https://en.cppreference.com/w/cpp/memory/shared_ptr/atomic2#:~:text=atomic%3Cshared_ptr%3CT%3E%3E%3A%3Await) that

> (atomic<shared_ptr<T>>::wait) compares load(order) with old and if they are equivalent then blocks ...

> Notes: two shared_ptrs are equivalent if they store the same pointer and either share ownership or are both empty.

I thought that `sp1` and `sp3` do not have the same control block, so they do not share ownership. Right? So the `wait` call should return immediately.

But by checking the source code, it seems that the STL implementation for `std::atomic<std::shared_ptr>::wait` compares the object pointer (the `char *`) only. It does not compare the control block pointer at all.

Is this a bug in STL? Or what am I missing here?

**Comment by fsb4000** [comment, marked as answer]
Thanks, it looks like a bug for me. https://eel.is/c++draft/util.smartptr.atomic.shared#lib:wait,atomic%3cshared_ptr%3cT%3e%3e
```
void wait(shared_ptr<T> old, memory_order order = memory_order::seq_cst) const noexcept;

Preconditions: order is neither memory_­order​::​release nor memory_­order​::​acq_­rel.

Effects: Repeatedly performs the following steps, in order:
    Evaluates load(order) and compares it to old.
    If the two are not equivalent, returns.
    Blocks until it is unblocked by an atomic notifying operation or is unblocked spuriously.

Remarks: Two shared_­ptr objects are equivalent if they store the same pointer and either share ownership or are both empty.
```

> returns successfully under gcc/clang

By the way LLVM libc++ doesn't implement `Atomic std::shared_ptr and std::weak_ptr` for now: https://godbolt.org/z/vKTE5s6d6

**Reply by zwhfly** [reply]
Oh I didn't know that godbolt.org uses libstdc++ with clang by default. Have edited the original post.
